build.gradle.kts requires explicitly java11

This requires the presence of exactly `jdk11`, which is becoming less and less likely as it was EOLed, let's at least use 17.

I have a somewhat similar issue. 

Platform: MacOS 14.2.1 (aarch64)
```sh
❯ java -version
openjdk version "21.0.1" 2023-10-17
OpenJDK Runtime Environment Homebrew (build 21.0.1)
OpenJDK 64-Bit Server VM Homebrew (build 21.0.1, mixed mode, sharing)
```

```sh
❯ /usr/libexec/java_home
/opt/homebrew/Cellar/openjdk/21.0.1/libexec/openjdk.jdk/Contents/Home
```

On a newly initialized gradle project (created with gradle 8.5). I can query installed toolchains
```sh
❯ gradle -q javaToolchains

 + Options
     | Auto-detection:     Enabled
     | Auto-download:      Enabled

 + Amazon Corretto JDK 20.0.2.1+10-FR
     | Location:           /Users/asoli/Library/Java/JavaVirtualMachines/corretto-20.0.2.1/Contents/Home
     | Language Version:   20
     | Vendor:             Amazon Corretto
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        IntelliJ IDEA

 + Homebrew JDK 21.0.1
     | Location:           /opt/homebrew/Cellar/openjdk/21.0.1/libexec/openjdk.jdk/Contents/Home
     | Language Version:   21
     | Vendor:             Homebrew
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        Current JVM
```

When running `./gradlew` (even if I update our gradle version from 8.4 to 8.5) I get this error:
```sh
❯ ./gradlew build

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'restate-e2e'.
> Failed to calculate the value of task ':compileJava' property 'javaCompiler'.
   > No matching toolchains found for requested specification: {languageVersion=11, vendor=any, implementation=vendor-specific} for MAC_OS on aarch64.
      > No locally installed toolchains match and toolchain download repositories have not been configured.

* Try:
> Learn more about toolchain auto-detection at https://docs.gradle.org/8.4/userguide/toolchains.html#sec:auto_detection.
> Learn more about toolchain repositories at https://docs.gradle.org/8.4/userguide/toolchains.html#sub:download_repositories.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 524ms
```

--- a/.github/workflows/e2e.yaml
+++ b/.github/workflows/e2e.yaml
@@ -163,7 +163,7 @@ jobs:
           RESTATE_CONTAINER_IMAGE: ${{ inputs.restateCommit != '' && 'localhost:5000/restatedev/restate:latest' || '' }}
           JAVA_SDK_LOCAL_BUILD: ${{ inputs.sdkJavaCommit != '' && 'true' || '' }}
         with:
-          arguments: -Djib.console=plain check
+          arguments: --continue -Djib.console=plain check
           build-root-directory: e2e
 
       # Upload container logs

--- a/.github/workflows/e2e.yaml
+++ b/.github/workflows/e2e.yaml
@@ -83,7 +83,7 @@ jobs:
       - uses: actions/setup-java@v3
         with:
           distribution: 'temurin'
-          java-version: '11'
+          java-version: '17'
 
       # Setup node
       - uses: actions/setup-node@v3

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -13,10 +13,14 @@ import org.gradle.nativeplatform.platform.internal.DefaultNativePlatform.getCurr
 
 plugins {
   java
-  alias(libs.plugins.spotless)
+
+  // Declare plugins used by subprojects
+  id("com.diffplug.spotless") version "6.22.0" apply false
   kotlin("jvm") version "1.9.22" apply false
   kotlin("plugin.serialization") version "1.9.22" apply false
-  alias(libs.plugins.jib) apply false
+  id("com.google.cloud.tools.jib") version "3.2.1" apply false
+  id("net.ltgt.errorprone") version "3.1.0" apply false
+  id("com.google.protobuf") version "0.9.4" apply false
 }
 
 val restateVersion = libs.versions.restate.get()

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -14,8 +14,8 @@ import org.gradle.nativeplatform.platform.internal.DefaultNativePlatform.getCurr
 plugins {
   java
   alias(libs.plugins.spotless)
-  kotlin("jvm") version "1.8.21" apply false
-  kotlin("plugin.serialization") version "1.8.21" apply false
+  kotlin("jvm") version "1.9.22" apply false
+  kotlin("plugin.serialization") version "1.9.22" apply false
   alias(libs.plugins.jib) apply false
 }
 

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -75,7 +75,10 @@ allprojects {
     }
   }
 
-  java { toolchain { languageVersion = JavaLanguageVersion.of(11) } }
+  java {
+    sourceCompatibility = JavaVersion.VERSION_17
+    targetCompatibility = JavaVersion.VERSION_17
+  }
 }
 
 buildscript {

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -95,15 +95,8 @@ subprojects {
   apply(plugin = "kotlin")
   apply(plugin = "com.diffplug.spotless")
 
-  val testReport =
-      tasks.register<TestReport>("testReport") {
-        destinationDirectory.set(layout.buildDirectory.dir("reports/tests/test"))
-        testResults.setFrom(subprojects.mapNotNull { it.tasks.findByPath("test") })
-      }
-
   tasks.withType<Test> {
     useJUnitPlatform()
-    finalizedBy(testReport)
     testLogging {
       events(
           TestLogEvent.PASSED,

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -79,10 +79,8 @@ allprojects {
     }
   }
 
-  java {
-    sourceCompatibility = JavaVersion.VERSION_17
-    targetCompatibility = JavaVersion.VERSION_17
-  }
+  // Configure the java toolchain to use. If not found, it will be downloaded automatically
+  java { toolchain { languageVersion = JavaLanguageVersion.of(11) } }
 }
 
 buildscript {

--- a/contracts/build.gradle.kts
+++ b/contracts/build.gradle.kts
@@ -13,7 +13,7 @@ import com.google.protobuf.gradle.protobuf
 plugins {
   java
   kotlin("jvm")
-  alias(libs.plugins.protobuf)
+  id("com.google.protobuf")
 }
 
 dependencies {

--- a/services/http-server/build.gradle.kts
+++ b/services/http-server/build.gradle.kts
@@ -9,10 +9,8 @@
 
 plugins {
   java
-  idea
-  alias(libs.plugins.shadowJar)
-  alias(libs.plugins.errorprone)
-  alias(libs.plugins.jib)
+  id("net.ltgt.errorprone")
+  id("com.google.cloud.tools.jib")
 }
 
 dependencies {
@@ -43,7 +41,6 @@ jib {
   }
 }
 
-// Use gradle shadowJar to build the fat jar
 tasks.jar {
   manifest { attributes["Main-Class"] = "dev.restate.e2e.testing.externalhttpserver.Main" }
 }

--- a/services/java-services/build.gradle.kts
+++ b/services/java-services/build.gradle.kts
@@ -13,9 +13,8 @@ import com.google.protobuf.gradle.protobuf
 plugins {
   java
   application
-  alias(libs.plugins.protobuf)
-  alias(libs.plugins.shadowJar)
-  alias(libs.plugins.jib)
+  id("com.google.protobuf")
+  id("com.google.cloud.tools.jib")
 }
 
 dependencies {
@@ -65,7 +64,6 @@ jib {
   }
 }
 
-// Use gradle shadowJar to build the fat jar
 tasks.jar { manifest { attributes["Main-Class"] = "dev.restate.e2e.services.Main" } }
 
 application { mainClass.set("dev.restate.e2e.services.Main") }

--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -80,14 +80,6 @@ dependencyResolutionManagement {
       library("awaitility", "org.awaitility", "awaitility-kotlin").versionRef("awaitility")
 
       library("errorprone", "com.google.errorprone", "error_prone_core").versionRef("errorprone")
-
-      // Plugins
-      plugin("spotless", "com.diffplug.spotless").version("6.22.0")
-      plugin("protobuf", "com.google.protobuf").version("0.9.4")
-      plugin("test-logger", "com.adarshr.test-logger").version("4.0.0")
-      plugin("errorprone", "net.ltgt.errorprone").version("3.1.0")
-      plugin("shadowJar", "com.github.johnrengelman.shadow").version("7.1.2")
-      plugin("jib", "com.google.cloud.tools.jib").version("3.2.1")
     }
   }
 }

--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -9,6 +9,8 @@
 
 rootProject.name = "restate-e2e"
 
+plugins { id("org.gradle.toolchains.foojay-resolver-convention") version "0.7.0" }
+
 include(
     "contracts",
     "services:java-services",

--- a/tests/build.gradle.kts
+++ b/tests/build.gradle.kts
@@ -73,66 +73,71 @@ tasks {
     }
   }
 
-  register<Test>("testAlwaysSuspending") {
-    environment =
-        environment +
-            baseRestateEnvironment(name) +
-            mapOf("RESTATE_WORKER__INVOKER__INACTIVITY_TIMEOUT" to "0s")
+  val testAlwaysSuspending by
+      registering(Test::class) {
+        environment =
+            environment +
+                baseRestateEnvironment(name) +
+                mapOf("RESTATE_WORKER__INVOKER__INACTIVITY_TIMEOUT" to "0s")
 
-    useJUnitPlatform {
-      // Run all the tests with always-suspending or only-always-suspending tag
-      includeTags("always-suspending | only-always-suspending")
-    }
-    // Increase a bit the default timeout
-    systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "30 s"
-  }
+        useJUnitPlatform {
+          // Run all the tests with always-suspending or only-always-suspending tag
+          includeTags("always-suspending | only-always-suspending")
+        }
+        // Increase a bit the default timeout
+        systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "30 s"
+      }
 
-  register<Test>("testSingleThreadSinglePartition") {
-    environment =
-        environment +
-            baseRestateEnvironment(name) +
-            mapOf(
-                "RESTATE_WORKER__PARTITIONS" to "1",
-                "RESTATE_TOKIO_RUNTIME__WORKER_THREADS" to "1",
-                "RESTATE_TOKIO_RUNTIME__MAX_BLOCKING_THREADS" to "1",
-            )
+  val testSingleThreadSinglePartition by
+      registering(Test::class) {
+        environment =
+            environment +
+                baseRestateEnvironment(name) +
+                mapOf(
+                    "RESTATE_WORKER__PARTITIONS" to "1",
+                    "RESTATE_TOKIO_RUNTIME__WORKER_THREADS" to "1",
+                    "RESTATE_TOKIO_RUNTIME__MAX_BLOCKING_THREADS" to "1",
+                )
 
-    useJUnitPlatform {
-      // Run all the tests with either no tags, or always-suspending tag
-      includeTags("none() | always-suspending")
-    }
-    // Increase a bit the default timeout
-    systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "30 s"
-  }
+        useJUnitPlatform {
+          // Run all the tests with either no tags, or always-suspending tag
+          includeTags("none() | always-suspending")
+        }
+        // Increase a bit the default timeout
+        systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "30 s"
+      }
 
-  register<Test>("testLazyState") {
-    environment =
-        environment +
-            baseRestateEnvironment(name) +
-            mapOf(
-                "RESTATE_WORKER__INVOKER__DISABLE_EAGER_STATE" to "true",
-            )
+  val testLazyState by
+      registering(Test::class) {
+        environment =
+            environment +
+                baseRestateEnvironment(name) +
+                mapOf(
+                    "RESTATE_WORKER__INVOKER__DISABLE_EAGER_STATE" to "true",
+                )
 
-    useJUnitPlatform {
-      // Run all the tests with either no tags, or always-suspending tag
-      includeTags("lazy-state")
-    }
-  }
+        useJUnitPlatform {
+          // Run all the tests with either no tags, or always-suspending tag
+          includeTags("lazy-state")
+        }
+      }
 
-  register<Test>("testPersistedTimers") {
-    environment =
-        environment +
-            baseRestateEnvironment(name) +
-            mapOf("RESTATE_WORKER__TIMERS__NUM_TIMERS_IN_MEMORY_LIMIT" to "1")
+  val testPersistedTimers by
+      registering(Test::class) {
+        environment =
+            environment +
+                baseRestateEnvironment(name) +
+                mapOf("RESTATE_WORKER__TIMERS__NUM_TIMERS_IN_MEMORY_LIMIT" to "1")
 
-    useJUnitPlatform {
-      // Run all the tests with always-suspending or only-always-suspending tag
-      includeTags("timers")
-    }
-    // Increase a bit the default timeout
-    systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "20 s"
-  }
+        useJUnitPlatform {
+          // Run all the tests with always-suspending or only-always-suspending tag
+          includeTags("timers")
+        }
+        // Increase a bit the default timeout
+        systemProperties["junit.jupiter.execution.timeout.testable.method.default"] = "20 s"
+      }
 
+  // Common configuration for all test tasks
   withType<Test>().configureEach {
     dependsOn(":services:http-server:jibDockerBuild")
     dependsOn(":services:java-services:jibDockerBuild")
@@ -140,11 +145,11 @@ tasks {
 
     maxParallelForks = 3
   }
-}
 
-tasks.named("check") {
-  dependsOn("testAlwaysSuspending")
-  dependsOn("testSingleThreadSinglePartition")
-  dependsOn("testPersistedTimers")
-  dependsOn("testLazyState")
+  check {
+    dependsOn(testAlwaysSuspending)
+    dependsOn(testSingleThreadSinglePartition)
+    dependsOn(testPersistedTimers)
+    dependsOn(testLazyState)
+  }
 }

